File: Design-Patterns/Visitor/CSharp/Equipments/Armor.cs
using Visitor.Visitors;

namespace Visitor.Equipments
{
    public class Armor : Equipment
    {
        public int Endurance { get; protected set; }

        public Armor(string name, int loading, int endurance) : base(name, loading)
        {
            Endurance = endurance;
        }

        public override void RunVisitor(EquipmentVisitor visitor)
        {
            visitor.VisitArmor(this);
        }

        public void TakeDamage()
        {
            Endurance -= 1;
            if (Endurance < 0) Endurance = 0;
        }
    }
}

File: Design-Patterns/Visitor/Cpp/armor.cpp
#include "equipment.h"
#include "equipment_visitor.h"

armor::armor(std::string name, int loading, int endurance) : equipment(name, loading)
{
	endurance_ = endurance;
}

int armor::get_endurance()
{
	return endurance_;
}


void armor::run_visitor(std::shared_ptr<equipment_visitor> visitor)
{
	visitor->visit_armor(std::shared_ptr<armor>(std::shared_ptr<armor>(), this));
}

void armor::take_damage()
{
	endurance_ -= 1;
	if (endurance_ < 0) endurance_ = 0;
}
